internal/replication/blockreplicator_3node_test.go

```diff
@@ -0,0 +1,313 @@
+	leaderIndex1 := env.FindLeaderIndex()
+	require.True(t, leaderIndex1 >= 0)
+	t.Logf("Leader #1 index: %d", leaderIndex1)
+	err := env.nodes[leaderIndex1].Close()
+	require.NoError(t, err)
+	isLeaderCond := func() bool {
+		return env.FindLeaderIndex() >= 0
+	}
+	assert.Eventually(t, isLeaderCond, 30*time.Second, 100*time.Millisecond)
+
+	//close the 2nd leader, remaining node has no leader
+	leaderIndex2 := env.FindLeaderIndex()
+	require.True(t, leaderIndex2 >= 0)
+	t.Logf("Leader #2 index: %d", leaderIndex2)
+	err = env.nodes[leaderIndex2].Close()
+	require.NoError(t, err)
+
+	//remaining node has no leader
+	var lastIndex int
+	for i, _ := range env.nodes {
+		if i == leaderIndex1 || i == leaderIndex2 {
+			continue
+		}
+		lastIndex = i
+	}
+	noLeaderCond := func() bool {
+		err := env.nodes[lastIndex].blockReplicator.IsLeader()
+		if err != nil && err.LeaderID == 0 {
+			return true
+		}
+		return false
+	}
+	assert.Eventually(t, noLeaderCond, 30*time.Second, 100*time.Millisecond)
+
+	err = env.nodes[lastIndex].Close()
+	require.NoError(t, err)
+}
+
+// Scenario:
+// - Start 1 node, no leader
+// - Start 2nd node, wait for leader,
+// - Start 3rd, wait for consistent leader on all
+func TestBlockReplicator_3Node_StartStepClose(t *testing.T) {
+	env := createClusterEnv(t, "info", 3)
+	require.Equal(t, 3, len(env.nodes))
+	defer os.RemoveAll(env.testDir)
+
+	err := env.nodes[0].Start()
+	require.NoError(t, err)
+	noLeaderCond := func() bool {
+		err := env.nodes[0].blockReplicator.IsLeader()
+		if err != nil && err.LeaderID == 0 {
+			return true
+		}
+		return false
+	}
+	assert.Eventually(t, noLeaderCond, 30*time.Second, 100*time.Millisecond)
+
+	err = env.nodes[1].Start()
+	require.NoError(t, err)
+	isLeaderCond1 := func() bool {
+		return env.AgreedLeaderIndex(0, 1) >= 0
+	}
+	assert.Eventually(t, isLeaderCond1, 30*time.Second, 100*time.Millisecond)
+
+	err = env.nodes[2].Start()
+	require.NoError(t, err)
+	isLeaderCond2 := func() bool {
+		return env.AgreedLeaderIndex() >= 0
+	}
+	assert.Eventually(t, isLeaderCond2, 30*time.Second, 100*time.Millisecond)
+
+	//close all
+	for _, node := range env.nodes {
+		err := node.Close()
+		require.NoError(t, err)
+	}
+}
+
+// Scenario:
+// - Start 3 nodes together, wait for leader,
+// - Stop the leader, wait for new leader,
+// - Restart the node, wait for consistent leader on all
+// - Stop a follower, wait for new leader,
+// - Restart the node, wait for consistent leader on all
+func TestBlockReplicator_3Node_Restart(t *testing.T) {
+	env := createClusterEnv(t, "info", 3)
+	require.Equal(t, 3, len(env.nodes))
+	defer os.RemoveAll(env.testDir)
+
+	// start 3 at once
+	for _, node := range env.nodes {
+		err := node.Start()
+		require.NoError(t, err)
+	}
+
+	// wait for an agreed leader
+	sameLeaderCond := func() bool {
+		return env.AgreedLeaderIndex() >= 0
+	}
+	assert.Eventually(t, sameLeaderCond, 30*time.Second, 100*time.Millisecond)
```

Comment:
as it is repetitive, we can make it into a function and reuse everywhere.

Reply:
done